### DISCUSSION TITLE
How do I clear map from previous style applied

### DISCUSSION THREAD

**Discussion by daniluk4000** [openlayers/ol-mapbox-style#1194] [Q&A] [answered]
Hey all! I don't get it. 

I try to apply Carto basemap styles to map, then change it on theme change. 

I have tried:
- Using applyStyle with removeLayer on theme change
- Using apply with layer group change
- Using MapboxVectorLayer and then disposing it & removing from map

No matter how I try, I get this on theme change

![image](https://github.com/user-attachments/assets/7f86ff49-e990-4e77-ae76-a2e486542d92)

It seems like source is cached... somewhere inside ol-mapbox-style.

Is there something we can do to apply completely different style to map? Why is even like this?

Thanks!

**Comment by daniluk4000** [comment]
Example style url: https://basemaps.cartocdn.com/gl/dark-matter-gl-style/style.json

I have tried to download it and change source to random name, but that didn't help as well :(

Seems like cache lives in util.js, but we have no control over it

**Comment by ahocevar** [comment, marked as answer]
The problem is that https://basemaps.cartocdn.com/gl/dark-matter-gl-style/style.json has
```json
  "id": "voyager"
```
and https://basemaps.cartocdn.com/gl/voyager-gl-style/style.json also has
```json
  "id": "voyager"
```
ol-mapbox-style caches styles by id to allow for efficient theme switching. If both themes have the same id, that won't work. A workaround is to set the id to the style's url:
```js
  group = new LayerGroup();
  const url = themes[currentTheme];
  fetch(url)
    .then((r) => r.json())
    .then((json) => {
      json.id = url;
      apply(group, json);
      map.addLayer(group);
    });
```
Here is a working codesandbox with theme switching: https://codesandbox.io/p/sandbox/8nsy64

**Reply by daniluk4000** [reply]
So I was changing wrong thing then. Thank you!
